### ISSUE TITLE
DRM content fails to play sometimes on Android

### ISSUE THREAD

**Issue by efkann** [TheWidlarzGroup/react-native-video#3082]
# Bug

Android Exoplayer fails to play DRM protected content sometimes. This bug is hard to explain, but I'll try my best.

This rarely happens on a bare react native project. But when it comes to expo, you use a expo development client and the content plays on the first try but never works again until you stop and start the dev client.

## Platform
Android

Which player are you experiencing the problem on:

Android Exoplayer

Library version: 6.0.0-alpha.5

## Steps To Reproduce

Think about two screens. 

-HomeScreen
-PlayerScreen

Try to navigate to Player screen from Home screen and go back to Home screen and then again navigate to Player screen. Repeat this 20-25 times and it will fail to play the protected content in one of those tries at least, in bare workflow. But in Expo, it plays the content at first navigation to Player screen but when you go back to Home screen and navigate back to the Player screen, it fails to play it. And it will never be able to play it until you go and restart the dev client.

Actually, the content looks like its playing and you can see the progress, play or pause but the screen is all black. I believe, player thinks its clear content but actually its not so it starts playing it but you cant see the content which also explains the fact that there's not even a single error printed out.

See the `else if` block I'm talking about below.

## Expected behaviour

It should play the protected content 

## Reproducible sample code

Since this issue mostly happens on Expo and rarely happens on bare workflow. I'm not sure what to add here. But I've been debugging this for a while and I might have an idea why this happens.

For testing DRM, you can try these :

sourceUrl: https://storage.googleapis.com/shaka-demo-assets/angel-one-widevine/dash.mpd
licenseUrl: https://cwip-shaka-proxy.appspot.com/no_auth

https://github.com/react-native-video/react-native-video/blob/05864dc64a588f9ffeeded45855eb390fd15b1d8/android/src/main/java/com/brentvatne/exoplayer/ReactExoplayerView.java#L553-L593

At the bottom of the block of code above I don't understand what this `else if` bit is doing actually.

```java
else if (srcUri != null) { 
     initializePlayerSource(self, null); 
 } 
```

Why we're passing null for `drmSessionManager` when `playerNeedsSource` is false ?

I also realized that, `playerNeedsSource` stays false for some reason and I believe it is the reason why the player tries to play DRM content but fails to do so. 

It's very interesting that for my Expo app, I prebuilt the app for Android and removed the`else if` block above and everything just worked perfectly.

## Note

I use Expo Router, but it's built on top of React Navigation so I dont think this is a bug related to navigation.

I'd be happy to provide if you need more information.

**Comment by efkann** [comment]
Also, I just noticed that, removing this if block that triggers reloadSource solves the problem as well.

https://github.com/react-native-video/react-native-video/blob/05864dc64a588f9ffeeded45855eb390fd15b1d8/android/src/main/java/com/brentvatne/exoplayer/ReactExoplayerView.java#L1479-L1481

Even when the player works perfectly, this if block runs the reloadSource which causes the execution of initializePlayer again. Is that necessary when the player loaded with a source first time ? 


But again, clear content works without an issue when using the library for both Expo and bare workflow.

**Comment by freeboub** [comment]
@efkann sorry for delayed answer, but it seems you are pointing the good code.
Did you find a fix for the issue ?
I confirm 

```
else if (srcUri != null) { 
     initializePlayerSource(self, null); 
 } 
```

Will cause a DRM issue !
BTW, the multiple thread handling in the if looks ugly to me :/ 
But I am not sure on how to reproduce the issue ...  I will go deeper in investigations to check if I can reproduce something

**Comment by efkann** [comment]
> @efkann sorry for delayed answer, but it seems you are pointing the good code. Did you find a fix for the issue ? I confirm
> 
> ```
> else if (srcUri != null) { 
>      initializePlayerSource(self, null); 
>  } 
> ```
> 
> Will cause a DRM issue ! BTW, the multiple thread handling in the if looks ugly to me :/ But I am not sure on how to reproduce the issue ... I will go deeper in investigations to check if I can reproduce something

You're probably right. This was a long time ago but as far as I remember, I couldn't find the exact code that causes the issue. 

I've been folllowing the recent changes with `expo-video` and will switch to it as it supports DRM now for both Android and iOS.

https://github.com/expo/expo/pull/26465

**Comment by freeboub** [comment]
Linked Pr merged, shall be fixed in 6.2.0
